### PR TITLE
do not cancel tuples if already done or failed

### PR DESCRIPTION
**Pull request by inelgnu**
related tests https://github.com/SubstraFoundation/substra-tests/pull/40

### CODE REVIEW

File: chaincode/testtuple.go [hunk 1]
```diff
@@ -468,6 +468,13 @@ func (testtuple *Testtuple) commitStatusUpdate(db *LedgerDB, testtupleKey string
 		return fmt.Errorf("update testtuple %s failed: %s", testtupleKey, err.Error())
 	}
 
+	// do not update if previous status is already Done or Failed
+	if StatusCanceled == newStatus {
+		if stringInSlice(testtuple.Status, []string{StatusDone, StatusFailed}) {
+			return nil
+		}
```
Comment (samlesu) [inline chaincode/testtuple.go]: What about also not update a `doing` tuple: as we cannot kill a running task we could let it finish and update its status accordingly?

Comment (inelgnu) [reply chaincode/testtuple.go]: done

File: chaincode/testtuple.go [hunk 2]
```diff
@@ -468,6 +468,13 @@ func (testtuple *Testtuple) commitStatusUpdate(db *LedgerDB, testtupleKey string
 		return fmt.Errorf("update testtuple %s failed: %s", testtupleKey, err.Error())
 	}
 
+	// do not update if previous status is already Done, Failed or Doing
+	if StatusCanceled == newStatus {
+		if stringInSlice(testtuple.Status, []string{StatusDone, StatusFailed, StatusDoing}) {
```
Comment (thibaultrobert) [inline chaincode/testtuple.go:473]: Not sure if the compilation optimize it or not but I would think it's better to use only one `if` statement with `&&`.

File: chaincode/traintuple.go [hunk 3]
```diff
@@ -613,6 +613,13 @@ func (traintuple *Traintuple) commitStatusUpdate(db *LedgerDB, traintupleKey str
 		return nil
 	}
 
+	// do not update if previous status is already Done, Failed or Doing
+	if StatusCanceled == newStatus {
+		if stringInSlice(traintuple.Status, []string{StatusDone, StatusFailed, StatusDoing}) {
+			return nil
+		}
+	}
+
```
Comment (thibaultrobert) [inline chaincode/traintuple.go:622]: Same as above.

File: chaincode/traintuple_composite.go [hunk 4]
```diff
@@ -567,6 +567,13 @@ func (traintuple *CompositeTraintuple) commitStatusUpdate(db *LedgerDB, traintup
 		return nil
 	}
 
+	// do not update if previous status is already Done, Failed or Doing
+	if StatusCanceled == newStatus {
+		if stringInSlice(traintuple.Status, []string{StatusDone, StatusFailed, StatusDoing}) {
+			return nil
+		}
+	}
+
```
Comment (thibaultrobert) [inline chaincode/traintuple_composite.go:576]: Same as above.

File: chaincode/tuple_aggregate.go [hunk 5]
```diff
@@ -522,6 +522,13 @@ func (tuple *Aggregatetuple) commitStatusUpdate(db *LedgerDB, aggregatetupleKey
 		return nil
 	}
 
+	// do not update if previous status is already Done, Failed or Doing
+	if StatusCanceled == newStatus {
+		if stringInSlice(tuple.Status, []string{StatusDone, StatusFailed, StatusDoing}) {
+			return nil
+		}
+	}
+
```
Comment (thibaultrobert) [inline chaincode/tuple_aggregate.go:531]: Same as above.

**Review by thibaultrobert — APPROVED** [review]
Some minor coding style, nothing problematic. I leave it up to you if you want to merge as it or if you update it before.

**Comment by inelgnu** [comment]
I merged for now, we will refactor this code soon

**Pull request merged**